File: demo/src/main/java/com/example/demo/entity/Product.java
package com.example.demo.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.Data;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
@Table(name = "products")
public class Product extends Base<String> implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull(message = "Product name cannot be null")
    @Column(nullable = false)
    private String name;

    private String description;

    @Min(value = 0, message = "Import price must be greater than or equal to 0")
    private Double importPrice;

    @Min(value = 0, message = "List price must be greater than or equal to 0")
    private Double listPrice;

    @Min(value = 0, message = "Sell price must be greater than or equal to 0")
    private Double sellPrice;

    @Min(value = 0, message = "Quantity must be greater than or equal to 0")
    private Double quantity;

    @Min(value = 0, message = "Sold quantity must be greater than or equal to 0")
    private Double soldQuantity;

    @ManyToMany(fetch = FetchType.LAZY, cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
    @JoinTable(
            name = "product_category",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<Category> categories = new HashSet<>();

    @JsonIgnore
    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Image> images = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Product)) return false;
        Product product = (Product) o;
        return id != null && id.equals(product.id);
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : 0;
    }
}

File: camerashop_fe/src/model/Product.ts
import Image from "./Image";
import ImageModel from "./Image";

class Product {
    id: number;
    name?: string;
    importPrice?: number | undefined;  // Thêm | undefined
    listPrice?: number | undefined;  // Thêm | undefined
    sellPrice?: number | undefined;  // Thêm | undefined
    description?: string;
    quantity?: number | undefined;  // Thêm | undefined
    soldQuantity?: number | undefined;  // Thêm | undefined
    categoryIds: number[];
    images: Image[];


    constructor(id: number, name: string, importPrice: number | undefined, listPrice: number | undefined, sellPrice: number | undefined, description: string, quantity: number | undefined, soldQuantity: number | undefined, categoryIds: number[], images: ImageModel[]) {
        this.id = id;
        this.name = name;
        this.importPrice = importPrice;
        this.listPrice = listPrice;
        this.sellPrice = sellPrice;
        this.description = description;
        this.quantity = quantity;
        this.soldQuantity = soldQuantity;
        this.categoryIds = categoryIds;
        this.images = images;
    }
}

export default Product;
